feat!: 4k collateral high performance masternode implementation

ogabrielides opened the pull request:
<!--
*** Please remove the following help text before submitting: ***

Provide a general summary of your changes in the Title above

Pull requests without a rationale and clear improvement may be closed
immediately.

Please provide clear motivation for your patch and explain how it improves
Dash Core user experience or Dash Core developer experience
significantly:

* Any test improvements or new tests that improve coverage are always welcome.
* All other changes should have accompanying unit tests (see `src/test/`) or
  functional tests (see `test/`). Contributors should note which tests cover
  modified code. If no tests exist for a region of modified code, new tests
  should accompany the change.
* Bug fixes are most welcome when they come with steps to reproduce or an
  explanation of the potential issue as well as reasoning for the way the bug
  was fixed.
* Features are welcome, but might be rejected due to design or scope issues.
  If a feature is based on a lot of dependencies, contributors should first
  consider building the system outside of Dash Core, if possible.
-->

## Issue being fixed or feature implemented
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->


## What was done?
<!--- Describe your changes in detail -->
Implementation of 4k collateral HPMN.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->


## Breaking Changes
<!--- Please describe any breaking changes your code introduces -->


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

thephez commented:
@PastaPastaPasta @UdjinM6 We need to make sure this doesn't create any issues with sentinel. Please consider that when reviewing. @nmarley You may know best about that.

UdjinM6 reviewed (changes requested):
👍

few notes

thephez reviewed (changes requested):
I see @UdjinM6 sneaked his review in before I posted mine so there might be some overlap :slightly_smiling_face:

ogabrielides commented on src/rpc/evo.cpp:
```diff
@@ -348,7 +367,7 @@ static void protx_register_fund_help(const JSONRPCRequest& request)
         },
         RPCExamples{
-            HelpExampleCli("protx", "register_fund \"XrVhS9LogauRJGJu2sHuryjhpuex4RNPSb\" \"1.2.3.4:1234\" \"Xt9AMWaYSz7tR7Uo7gzXA3m4QmeWgrR3rr\" \"93746e8731c57f87f79b3620a7982924e2931717d49540a85864bd543de11c43fb868fd63e501a1db37e19ed59ae6db4\" \"Xt9AMWaYSz7tR7Uo7gzXA3m4QmeWgrR3rr\" 0 \"XrVhS9LogauRJGJu2sHuryjhpuex4RNPSb\"")
+                HelpExampleCli("protx", "register_fund \"XrVhS9LogauRJGJu2sHuryjhpuex4RNPSb\" \"1.2.3.4:1234\" \"Xt9AMWaYSz7tR7Uo7gzXA3m4QmeWgrR3rr\" \"93746e8731c57f87f79b3620a7982924e2931717d49540a85864bd543de11c43fb868fd63e501a1db37e19ed59ae6db4\" \"Xt9AMWaYSz7tR7Uo7gzXA3m4QmeWgrR3rr\" 0 \"XrVhS9LogauRJGJu2sHuryjhpuex4RNPSb\"")
```
Sorry, not intentional.

ogabrielides commented on src/rpc/evo.cpp:
```diff
@@ -433,13 +452,14 @@ static void protx_register_submit_help(const JSONRPCRequest& request)
                                        const bool isPrepareRegister)
 {
-    if (isFundRegister && (request.fHelp || (request.params.size() < 7 || request.params.size() > 9))) {
+    if (isFundRegister && (request.fHelp || (request.params.size() < 8 || request.params.size() > 10))) {
```
Mistake

ogabrielides commented:
@UdjinM6 @UdjinM6 Applied suggestions.
Implemented also:
- `platformNodeID` null is not allowed anymore
- `platformNodeID` uniqueness in the network
- rebased

thephez reviewed (commented):
:notebook: Release notes ACK

nmarley commented:
@thephez I've done a _very light_ review, and I don't think this will cause any issues w/sentinel. Sentinel only deals w/proposal payments and not regular MN payouts, and I don't think this PR has touched the governance mechanism.

Even for PRs that do, as long as Core handles the SB creation appropriately, I don't think it would be an issue. Mostly the ball is on the Core side for that. Still, would be nice to get [sentinel logic ported into C++/Core](https://github.com/dashpay/sentinel/issues/77) if possible.

ogabrielides commented on src/governance/object.cpp:
```diff
@@ -626,7 +626,17 @@ int CGovernanceObject::CountMatchingVotes(vote_signal_enum_t eVoteSignalIn, vote
+                }
+            }
+            nCount += voteWeight;
         }
```
@nmarley Well Governance is changed here. HPMN have a 4x weighted vote.

nmarley replied:
@ogabrielides You're right! I must have been looking at a single commit instead of the entire thing, or I didn't see the scroll window for the rest of the files.

So any votes for proposals get weighted 4x then? Yes, that will definitely have an effect on Sentinel. Thanks for catching this.

We will need a way to identify HPMNs (via RPC) and 4x those votes, _or_ (my preference) have `getcurrentvotes` RPC call return another column for "vote weight", and return 1 for regular and 4 for HPMNs.

Either way I can make the required changes in Sentinel, but would prefer that getcurrentvotes return a weight because that puts all the weight logic into Core.

thephez commented on doc/release-notes-5039.md:
```diff
@@ -0,0 +1,18 @@
+Updated RPCs
+--------
+
+The RPC's `gobject getcurrentvotes` reply is enriched with the weight of the vote. Possible values 1 or 4. Example: 
```
```suggestion
The RPC's `gobject getcurrentvotes` reply is enriched by adding the vote weight at the end of each line. Possible values are 1 or 4. Example: 
```

thephez reviewed (commented):
Re-ack for release notes

UdjinM6 commented:
pls see https://github.com/UdjinM6/dash/commits/pr5039 for some suggestions

ogabrielides commented:
@UdjinM6 Applied suggestions. Please check

UdjinM6 reviewed (dismissed):
LGTM, utACK

PastaPastaPasta commented on src/evo/deterministicmns.cpp:
```diff
@@ -181,7 +182,27 @@ CDeterministicMNCPtr CDeterministicMNList::GetMNPayee() const
     }
 
-    CDeterministicMNCPtr best;
+    // Starting from v19 and until v20 (Platform release), HPMN will be rewarded 4 blocks in a row
+    // No need to check if v19 is active, since HPMN ProRegTx are allowed only after v19 activation
+    // TODO: Skip this code once v20 is active
+    CDeterministicMNCPtr best = nullptr;
+    ForEachMNShared(true, [&](const CDeterministicMNCPtr& dmn) {
+        if (dmn->pdmnState->nLastPaidHeight == nHeight) {
+            // We found the last MN Payee.
+            // If the last payee is a HPMN, we need to check its consecutive payments and pay him again if nConsecutivePayments < 3
+            if (dmn->nType == CDeterministicMN::TYPE_HIGH_PERFORMANCE_MASTERNODE && dmn->pdmnState->nConsecutivePayments < 3) {
+                best = dmn;
+            }
+        }
+        return;
+    });
+
+    if (best)
+        return best;
+
+    // Note: If the last payee was a regular MN or if the payee is a HPMN that was removed from the mnList then that's fine.
+    // We can proceed with classic MN payee selection
+
```
maybe guard this with v19 check

PastaPastaPasta commented on src/evo/deterministicmns.cpp:
```diff
@@ -1074,6 +1160,110 @@ void CDeterministicMNManager::CleanupCache(int nHeight)
+    curMNList.SetHeight(Params().GetConsensus().DIP0003Height - 1);
+    curMNList.SetBlockHash(::ChainActive()[Params().GetConsensus().DIP0003Height - 1]->GetBlockHash());
+
+    for (int nHeight = Params().GetConsensus().DIP0003Height; nHeight <= ::ChainActive().Height(); nHeight++) {
```
could use irange here I think

PastaPastaPasta commented on src/evo/deterministicmns.cpp:
```diff
@@ -1101,6 +1291,40 @@ static bool CheckService(const ProTx& proTx, CValidationState& state)
+        if (proTx.platformP2PPort != mainnetPlatformP2PPort) {
+            return state.Invalid(ValidationInvalidReason::TX_BAD_SPECIAL, false, REJECT_INVALID, "bad-protx-platform-p2p-port");
+        }
+    } else if (proTx.platformP2PPort == mainnetPlatformP2PPort || proTx.platformP2PPort < 1 || proTx.platformP2PPort > std::numeric_limits<uint16_t>::max()) {
```
we don't do a numeric limits for the normal p2p port right? This check is not needed as the type is uint16_t, it'll always be true

PastaPastaPasta commented on src/evo/deterministicmns.h:
```diff
@@ -40,18 +40,24 @@ class CDeterministicMN
     uint64_t internalId{std::numeric_limits<uint64_t>::max()};
 
 public:
+    static constexpr uint16_t TYPE_REGULAR_MASTERNODE = 0;
+    static constexpr uint16_t TYPE_HIGH_PERFORMANCE_MASTERNODE = 1;
```
make this an enum (I think not enum class here, as enum class cannot easily be converted to underlying type

PastaPastaPasta commented on src/evo/deterministicmns.h:
```diff
@@ -40,18 +40,24 @@ class CDeterministicMN
+
     CDeterministicMN() = delete; // no default constructor, must specify internalId
-    explicit CDeterministicMN(uint64_t _internalId) : internalId(_internalId)
+    explicit CDeterministicMN(uint64_t _internalId, bool highPerformanceMasternode = false) : internalId(_internalId)
```
I'd prefer this param be the above enum I suggested creating

ogabrielides replied:
@PastaPastaPasta No for this one since it is called here:
`std::make_shared<CDeterministicMN>(newList.GetTotalRegisteredCount(),proTx.nType==CProRegTx::TYPE_HIGH_PERFORMANCE_MASTERNODE);
`

I really don't want to mix CDeterministicMN enum for type and CProRegTx types.

PastaPastaPasta commented on src/evo/deterministicmns.h:
```diff
@@ -449,17 +479,28 @@ class CDeterministicMNListDiff
+            auto dmn = std::make_shared<CDeterministicMN>(mn);
+            addedMNs.push_back(dmn);
+        }
         tmp = ReadCompactSize(s);
```
this is a duplicate no?

ogabrielides replied:
@PastaPastaPasta Nope. it mn needs to be unserialized first with old format.

PastaPastaPasta commented on src/evo/dmnstate.cpp:
```diff
@@ -39,6 +39,7 @@ void CDeterministicMNState::ToJson(UniValue& obj) const
     obj.pushKV("service", addr.ToStringIPPort(false));
     obj.pushKV("registeredHeight", nRegisteredHeight);
     obj.pushKV("lastPaidHeight", nLastPaidHeight);
+    obj.pushKV("consecutivePayments", nConsecutivePayments);
```
haven't seen this get decreased on block rewind anywhere yet

ogabrielides replied:
@PastaPastaPasta Yup good catch my friend! 
It seems that @UdjinM6 missed that :)

thephez replied:
> @PastaPastaPasta Yup good catch my friend! It seems that @UdjinM6 missed that :)

Doesn't sound right :thinking: I bet it was just a test to see if @PastaPastaPasta was paying attention. :smile:

UdjinM6 replied:
`nConsecutivePayments` is updated in `BuildNewListFromBlock` and need no rollback logic I think

ogabrielides replied:
@PastaPastaPasta when we undo a block we just use another (previous) mnlist, there is nothing to undo in mnlist and/or mn states

PastaPastaPasta replied:
Gotcha, I wasn't sure if we needed any special logic, but I wanted to make sure we thought about it 👍

PastaPastaPasta commented on src/evo/dmnstate.h:
```diff
@@ -52,16 +99,42 @@ class CDeterministicMNState
     CScript scriptPayout;
     CScript scriptOperatorPayout;
 
+    uint160 platformNodeID{};
```
no validation on this?

ogabrielides replied:
@PastaPastaPasta What do you mean by validation here?

PastaPastaPasta replied:
Well, shouldn't the node be invalid for a null value? In order to become valid the operator will need to update service?

ogabrielides replied:
@PastaPastaPasta Not at this point (`CDeterministicMNState`). Validations are already made when check ProRegTx and ProUpServTx

PastaPastaPasta commented on src/evo/dmnstate.h:
```diff
@@ -52,16 +99,42 @@ class CDeterministicMNState
+            nPoSeBanHeight(s.nPoSeBanHeight),
+            nRegisteredHeight(s.nRegisteredHeight),
+            nLastPaidHeight(s.nLastPaidHeight),
+            nConsecutivePayments(0),
```
~not needed, ~ ~~well, why not copy this over?~~ oh, because it's old format. In that case, not needed

ogabrielides replied:
@PastaPastaPasta Of course it it is needed.
When migrating evoDB, first we read an `CDeterministicMNState_Oldformat` and then we construct `CDeterministicMNState` from `CDeterministicMNState_Oldformat`. So we need to copy all fields that are in common and set to default value the new fields.

PastaPastaPasta replied:
I specifically mean setting it to 0 isn't needed. B/c the default constructor / default value for the variable does it

ogabrielides replied:
Gotcha

PastaPastaPasta commented on src/evo/dmnstate.h:
```diff
@@ -52,16 +99,42 @@ class CDeterministicMNState
+            addr(s.addr),
+            scriptPayout(s.scriptPayout),
+            scriptOperatorPayout(s.scriptOperatorPayout),
+            platformNodeID(),
+            platformP2PPort(0),
+            platformHTTPPort(0){}
```
same not needed

ogabrielides replied:
Same

PastaPastaPasta commented on src/evo/dmnstate.h:
```diff
@@ -200,5 +285,4 @@ class CDeterministicMNStateDiff
     }
 };
 
-
```
not relevant

PastaPastaPasta commented on src/evo/providertx.h:
```diff
@@ -74,6 +79,13 @@ class CProRegTx
                 obj.scriptPayout,
                 obj.inputsHash
         );
+        if (obj.nVersion == BASIC_BLS_VERSION && obj.nType == TYPE_HIGH_PERFORMANCE_MASTERNODE) {
```
why bls check? oh, because this "version" also enables HPMNs, maybe that should be communicated in the version?

ogabrielides replied:
@PastaPastaPasta Yes, that's why. 
What do you mean by _communicated in the version_ ?

PastaPastaPasta replied:
something like `BASIC_BLS_VERSION` -> `BASIC_BLS_HPMN_COMPITABLE_VERSION` but that's a little verbose so idk 🤷

ogabrielides replied:
That's verbose I agree 🤷

PastaPastaPasta commented on src/evo/providertx.h:
```diff
@@ -115,15 +132,21 @@ class CProUpServTx
     static constexpr auto SPECIALTX_TYPE = TRANSACTION_PROVIDER_UPDATE_SERVICE;
     static constexpr uint16_t LEGACY_BLS_VERSION = 1;
     static constexpr uint16_t BASIC_BLS_VERSION = 2;
+    static constexpr uint16_t TYPE_REGULAR_MASTERNODE = 0;
+    static constexpr uint16_t TYPE_HIGH_PERFORMANCE_MASTERNODE = 1;
```
I hate having like 5 declarations of this

ogabrielides replied:
Me too by can't do much

PastaPastaPasta commented on src/evo/simplifiedmns.cpp:
```diff
@@ -68,6 +70,10 @@ void CSimplifiedMNListEntry::ToJson(UniValue& obj, bool extended) const
     obj.pushKV("nVersion", nVersion);
+    obj.pushKV("nType", nType);
+    if (nType == CDeterministicMN::TYPE_HIGH_PERFORMANCE_MASTERNODE) {
+        obj.pushKV("platformHTTPPort", platformHTTPPort);
```
why only platformhttp port?

ogabrielides replied:
@PastaPastaPasta As discussed with Platform team, only `platformHTTPPort` is included in SML entry.

PastaPastaPasta replied:
Do we have a justification for this in a document somewhere?

ogabrielides replied:
Yes I will send it to you.

PastaPastaPasta commented on src/evo/simplifiedmns.h:
```diff
@@ -32,6 +33,8 @@ class CSimplifiedMNListEntry
     CKeyID keyIDVoting;
     bool isValid{false};
+    uint16_t nType{CDeterministicMN::TYPE_REGULAR_MASTERNODE};
+    uint16_t platformHTTPPort{0};
```
why only the http?

ogabrielides replied:
@PastaPastaPasta As discussed with Platform team, only platformHTTPPort is included in SML entry.

HashEngineering replied:
@ogabrielides @PastaPastaPasta 

What about `platformNodeID` and `platformP2PPort` for the mobile clients?

Mobile clients use P2P for connectivity.  Additionally, we also have some support for masternode keys in our apps (iOS for now) and having the `platformNodeID` will also be useful.

HashEngineering replied:
Nevermind about `platformP2PPort` as the mobile clients will not be using this.

PastaPastaPasta commented on src/llmq/utils.cpp:
```diff
@@ -660,6 +660,11 @@ bool IsInstantSendLLMQTypeShared()
     return false;
 }
 
+bool IsPlatformLLMQType(Consensus::LLMQType llmqType)
+{
+    return Params().GetConsensus().llmqTypePlatform == llmqType;
+}
+
```
It'd be more readable in certain places if we added a wrapper / second name here something like `IsLLMQTypeHPMNOnly`

PastaPastaPasta commented on src/rpc/evo.cpp:
```diff
@@ -466,7 +486,8 @@ static UniValue protx_register_wrapper(const JSONRPCRequest& request,
     if (specific_legacy_bls_scheme)
         ptx.nVersion = CProRegTx::LEGACY_BLS_VERSION;
     else
-        ptx.nVersion = CProRegTx::GetVersion(llmq::utils::IsV19Active(::ChainActive().Tip()));
+        ptx.nVersion = CProRegTx::GetVersion(isV19active);
```
should be using brackets

ogabrielides replied:
@PastaPastaPasta Why? `CProRegTx::GetVersion` is a function

PastaPastaPasta replied:
the `if.. else..` should be using brackets so it's more clear what is inside the if and else scope. I know you're kinda not touching it, but you also introduced the `specific_legacy_bls_scheme` check so I can still yell at you for it :P

PastaPastaPasta commented on src/rpc/evo.cpp:
```diff
@@ -549,7 +570,293 @@ static UniValue protx_register_wrapper(const JSONRPCRequest& request,
         uint32_t collateralIndex = (uint32_t) -1;
         for (uint32_t i = 0; i < tx.vout.size(); i++) {
-            if (tx.vout[i].nValue == collateralAmount) {
+            if (tx.vout[i].nValue == CDeterministicMN::REGULAR_MASTERNODE_COLLATERAL) {
```
don't understand if this is correct, same with the static collateral value above

ogabrielides replied:
@PastaPastaPasta When calling `protx register_fund` and `register_fund_hpmn` we need to find 1000/4000 Dash in the vout so we can store collateralOutpoint.

PastaPastaPasta replied:
Well, I think I was more concerned with the duplication of code here more than anything

PastaPastaPasta commented on src/rpc/evo.cpp:
```diff
@@ -549,7 +570,293 @@ static UniValue protx_register_wrapper(const JSONRPCRequest& request,
+        keyIDVoting = ParsePubKeyIDFromAddress(request.params[paramIdx + 3].get_str(), "voting address");
+    }
+
+    int64_t operatorReward;
+    if (!ParseFixedPoint(request.params[paramIdx + 4].getValStr(), 2, &operatorReward)) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "operatorReward must be a number");
+    }
+    if (operatorReward < 0 || operatorReward > 10000) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "operatorReward must be between 0.00 and 100.00");
+    }
```
this is a bunch of duplicate logic

PastaPastaPasta commented on src/rpc/evo.cpp:
```diff
@@ -755,6 +1072,129 @@ static UniValue protx_update_service_legacy(const JSONRPCRequest& request)
+
+    CTxDestination feeSource;
+
+    // param feeSourceAddress
+    if (!request.params[7].isNull()) {
+        feeSource = DecodeDestination(request.params[7].get_str());
+        if (!IsValidDestination(feeSource))
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, std::string("Invalid Dash address: ") + request.params[7].get_str());
+    } else {
+        if (ptx.scriptOperatorPayout != CScript()) {
+            // use operator reward address as default source for fees
+            ExtractDestination(ptx.scriptOperatorPayout, feeSource);
+        } else {
+            // use payout address as default source for fees
+            ExtractDestination(dmn->pdmnState->scriptPayout, feeSource);
+        }
+    }
+
+    FundSpecialTx(wallet.get(), tx, ptx, feeSource);
+
+    SignSpecialTxPayloadByHash(tx, ptx, keyOperator);
+    SetTxPayload(tx, ptx);
+
+    return SignAndSendSpecialTx(request, tx);
```
likely all duplicate logic

PastaPastaPasta reviewed (commented):
reviewed basically all except the tests

PastaPastaPasta commented on src/governance/object.cpp:
```diff
@@ -619,14 +619,21 @@ bool CGovernanceObject::IsCollateralValid(std::string& strError, bool& fMissingC
+            auto dmn = mnList.GetMNByCollateral(votepair.first);
+            nCount += (dmn != nullptr && dmn->nType == CDeterministicMN::TYPE_HIGH_PERFORMANCE_MASTERNODE)
+                    ? CDeterministicMN::HIGH_PERFORMANCE_MASTERNODE_WEIGHT
+                    : CDeterministicMN::REGULAR_MASTERNODE_WEIGHT;
```
@UdjinM6 please make sure this shouldn't be a problematic change in logic

PastaPastaPasta commented on src/evo/deterministicmns.cpp:
```diff
@@ -1305,17 +1305,13 @@ static bool CheckPlatformFields(const ProTx& proTx, CValidationState& state)
         if (proTx.platformHTTPPort != mainnetPlatformHTTPPort) {
             return state.Invalid(ValidationInvalidReason::TX_BAD_SPECIAL, false, REJECT_INVALID, "bad-protx-platform-http-port");
         }
-    } else if (proTx.platformHTTPPort == mainnetPlatformHTTPPort || proTx.platformHTTPPort < 1 || proTx.platformHTTPPort > std::numeric_limits<uint16_t>::max()) {
```
don't we still want the check that it's not the main net httpport or nah?

ogabrielides replied:
@PastaPastaPasta Nope. Platform requested to relax this constraint.

ogabrielides commented:
@PastaPastaPasta @UdjinM6 Please review again.
Applied suggestions.

UdjinM6 reviewed (changes requested):
pls see https://github.com/UdjinM6/dash/commits/pr5039

PastaPastaPasta commented:
Please see https://github.com/dashpay/dash/commit/a2a4154dadfc04edaeebff5608d8182596271e9f

For some reason it breaks tests, which I cannot figure out at this moment, but conceptually, this is valuable imo

ogabrielides commented:
@PastaPastaPasta @UdjinM6 Applied suggestions (+fixed tests).
In addition, as discussed with @HashEngineering, `platformNodeID` is included in SML entries.
Please review again

PastaPastaPasta commented on src/governance/vote.cpp:
```diff
@@ -110,11 +111,15 @@ CGovernanceVote::CGovernanceVote(const COutPoint& outpointMasternodeIn, const ui
 
 std::string CGovernanceVote::ToString() const
 {
+    auto mnList = deterministicMNManager->GetListAtChainTip();
+    auto dmn = mnList.GetMNByCollateral(masternodeOutpoint);
+    int voteWeight = GetMnType(dmn->nType).voting_weight;
```
could be invalidated DMN

PastaPastaPasta commented on src/llmq/utils.h:
```diff
@@ -85,6 +85,8 @@ Consensus::LLMQType GetInstantSendLLMQType(bool deterministic);
 bool IsDIP0024Active(const CBlockIndex* pindex);
 bool IsV19Active(const CBlockIndex* pindex);
 const CBlockIndex* V19ActivationIndex(const CBlockIndex* pindex);
+static bool IsInstantSendLLMQTypeShared();
```
why is this new here?

ogabrielides replied:
Remove the line

PastaPastaPasta replied:
```suggestion
```

PastaPastaPasta commented on src/rpc/masternode.cpp:
```diff
@@ -666,7 +667,14 @@ static UniValue masternodelist(const JSONRPCRequest& request)
             objMN.pushKV("address", dmn.pdmnState->addr.ToString());
             objMN.pushKV("payee", payeeStr);
             objMN.pushKV("status", dmnToStatus(dmn));
+            objMN.pushKV("type", dmn.nType == MnType::HighPerformance.index ? "HighPerformance" : "Regular");
+            if (dmn.nType == MnType::HighPerformance.index) {
+                objMN.pushKV("platformNodeID", dmn.pdmnState->platformNodeID.ToString());
+                objMN.pushKV("platformP2PPort", dmn.pdmnState->platformP2PPort);
+                objMN.pushKV("platformHTTPPort", dmn.pdmnState->platformHTTPPort);
+            }
```
we should check for the nullptr state

PastaPastaPasta commented on src/evo/simplifiedmns.cpp:
```diff
@@ -80,6 +88,7 @@ void CSimplifiedMNListEntry::ToJson(UniValue& obj, bool extended) const
     }
 }
 
+// TODO: Invistigate if we can delete this constructor
```
added todo?

PastaPastaPasta reviewed (approved):
utACK for squash merge

knst commented on src/evo/deterministicmns.h:
```diff
@@ -62,30 +62,47 @@ class CDeterministicMN
     uint256 proTxHash;
     COutPoint collateralOutpoint;
     uint16_t nOperatorReward{0};
+    uint16_t nType{MnType::Regular.index};
```
why index is `uint8_t` but here is uint16_t? btw, better to use `enum class` here as more error prune (see other related comment)

knst commented on src/evo/dmn_types.h:
```diff
@@ -0,0 +1,41 @@
+class CDeterministicMNType
+{
+public:
+    uint8_t index;
```
may be change `uint8_t index` to `enum class MNType : uint8_t` ? 

Pros:
Firstly, switch inside `GetMnType()` and any similar code will show a compiler warning if any type is missing.
Secondly, instead "MnType::HighPerformance.index" you can write `MnType::HighPerformance`
Thirdly, you can remove .index from MnType
Cons:
Instead `(nType == MnType::HighPerformance.index)` need to write `nType == static_cast<int16_t>(MnType::HighPerformance)`

knst replied:
see #5200

knst commented on src/evo/deterministicmns.h:
```diff
@@ -40,18 +41,17 @@ class CDeterministicMN
+    explicit CDeterministicMN(uint64_t _internalId, bool highPerformanceMasternode = false) :
+        internalId(_internalId)
     {
+        highPerformanceMasternode ? nType = MnType::HighPerformance.index : nType = MnType::Regular.index;
```
nit: why not 
```
: internalId(_internalId),
  nType(highPerformanceMasternode ? HighPerformance.index : Regular.index)
{}
```

knst commented on src/evo/deterministicmns.cpp:
```diff
@@ -1013,7 +1117,10 @@ bool CDeterministicMNManager::IsProTxWithCollateral(const CTransactionRef& tx, u
     }
-    if (tx->vout[n].nValue != 1000 * COIN) {
+
+    CAmount expectedCollateral = GetMnType(proTx.nType).collat_amount;
```
Doesn't seems as expectedCollateral expected to be changed.
`const CAmount expectedCollateral = ...`

knst commented on src/evo/deterministicmns.cpp:
```diff
@@ -477,6 +514,13 @@ void CDeterministicMNList::UpdateMN(const CDeterministicMN& oldDmn, const std::s
+    if (dmn->nType == MnType::HighPerformance.index) {
+        if (!UpdateUniqueProperty(*dmn, oldState->platformNodeID, dmn->pdmnState->platformNodeID)) {
+            mnUniquePropertyMap = mnUniquePropertyMapSaved;
+            throw(std::runtime_error(strprintf("%s: Can't update a masternode %s with a duplicate platformNodeID=%s", __func__,
```
I though that () is not needed after throw. 

If remove them, the line will be more beautiful because no "))))" at the end of line but only ")))"

knst commented on src/evo/deterministicmns.cpp:
```diff
@@ -175,13 +177,35 @@ static bool CompareByLastPaid(const CDeterministicMN* _a, const CDeterministicMN
+            return;
+        });
+
+        if (best)
```
if it is single line condition IMO better to write `if (best) return best;` in line line rather than in 2 lines.
Otherwise add {}.

Because it is easy to make a mistake if edit code and remove only "if(best)" or remove only "return best" - the semantic may be changed significantly, but syntax still valid.

knst commented on src/evo/deterministicmns.cpp:
```diff
@@ -49,6 +50,7 @@ void CDeterministicMN::ToJson(UniValue& obj) const
     UniValue stateObj;
     pdmnState->ToJson(stateObj);
 
+    obj.pushKV("type", nType == MnType::HighPerformance.index ? "HighPerformance" : "Regular");
```
I see text constant is used at least twice: here and in rpc/masternode.cpp.

Instead better to create new const char pointer inside struct such as:
```
constexpr auto Regular = CDeterministicMNType{
    .voting_weight = 1,
    .collat_amount = 1000 * COIN,
    .descr = "Regular",
};
```


and here just write `obj.pushKV("type", getMnType(nType).desc)`

knst reviewed (commented):
some nits only. Not a blocker to merge! Can be done never or as spare PR someday.

ogabrielides commented:
> some nits only. Not a blocker to merge! Can be done never or as spare PR someday.

Sure. Can be applied in a separate PR. Thanks @knst !

UdjinM6 commented:
pls see https://github.com/UdjinM6/dash/commits/pr5039. also, I think expressions like `nType == MnType::HighPerformance.index` look pretty confusing. I like [the suggestion](https://github.com/dashpay/dash/pull/5039#discussion_r1104962589) made by @knst above, I would suggest to implement it here.

UdjinM6 reviewed (approved):
utACK

PastaPastaPasta reviewed (approved):
re-utACK for squash merge
